### PR TITLE
SNMP Trap Manager

### PR DESCRIPTION
**Pull request by samyscoub**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

### CODE REVIEW

**Comment by samyscoub** [comment]
I've just signed the CLA, can you recheck the PR please ?

Regards

**Comment by mattie47** [comment]
This looks really nice :-).

It might be worth adding a brief description of what this plugin actually allows you to do. 

I've skimmed through MIBUploader.md and done a quick google on SNMPTT.

From what I've gathered:

MIBUploader is a plugin which adds SNMP Trap functionality into LibreNMS through SNMPTT.

This allows LibreNMS alerts and rules to be generated based on traps received.

In section "MIB Uploading", can SNMPTT not simply pick up the mibs in the /mibs directory?

Or have I misunderstood the purpose of uploading mibs?

Thanks,

Matt

**Comment by laf** [comment]
@samyscoub It's not showing that you've signed it still, can you retry please?

Ok so this looks like a great addition but will need quite a bit of work as it really needs integrating in to the codebase rather than being a separate folder and using the plugin system (the plugin system will most likely be scrapped). Happy to help make this all happen. Can I ask though, is this all of your own work? We have to be careful of introducing code that has been written for observium.

The license is MIT which whilst compatible with our GPLv3 license, should really be licensed as GPLv3 to avoid any confusion. Can you re-license?

**Comment by samyscoub** [comment]
@mattie47 : currently, this module is only managing MIBs that are uploaded through the form into a specific SQL table. Of course, you can upload MIBs from the *mibs* directory of librenms, but the plugin does not ( in this release ) automatically search to translate, parse and save into database all MIBs that are present into the *mibs* directory ( used for metrology and devices information purpose ). This computing take some ressources, so make it automatic can have a performance cost.

@laf : this code has been made by my company for a big french client. This client and my company has choosen that we can donate this work to open source librenms code.  I've re-accepted the CLA and change Licence to GPLV3. Can you check that it is ok for you ?
No problem to make changes if you guide me to what do you desire.

**Comment by murrant** [comment]
This is great thanks for contributing.

I think this should either be better integrated into LibreNMS or published in a separate repository as a plugin.

If you want to integrate this into LibreNMS, here are some things needed off the top of my head:
1. Rename this to from mib uploader to custom traps or something else (the name mib uploader will cause confusion)
2. Move things out of the plugins folder into proper places
3. Dedicated menu for traps + this
4. Use autoloading instead of require

Additionally, if you want to update this independently, you could split part of this out into a php package and register with packagist.org  Then we could just pull that in as a dependency.

Just my thoughts, willing to work with you to decide the best course of action.

**Comment by samyscoub** [comment]
@murrant , no problem if you want to work with me on this feature.
First thing with the name, what do you thing of "Trap Manager" instead of "Mib uploader" ?
About the choice between separated plugin or more integreated feature, I have no opinion.
In your opinion, what is the better solution to easy maintain the feature ?
A core feature is always better for a product but even if it is a separated feature, it must be highlighted  into the documentation.

**Comment by laf** [comment]
I'd definitely go for core integration.

How about 'SNMP Trap Manager'?

**Comment by samyscoub** [comment]
@laf , I like your name proposal for this feature

**Comment by murrant** [comment]
Core integration would be preferred yes.  +1 for SNMP Trap Manager

**Comment by samyscoub** [comment]
Ok for the name. So how do we start @murrant  ?
I gave you access to the repository.
Can you start a list of tasks to make this core integration ?

**Comment by laf** [comment]
I'll try and go through it tonight.

**Comment by laf** [comment]
Quick high level pass for now:

Remove the LICENSE file

Why has the composer file been changed?

Remove mibuploader/AUTHORS mibuploader/ChangeLog mibuploader/LICENSE mibuploader/README.md

Move mibuploader/doc/MIBUploader.md to doc/Extensions/SNMP-Trap-Manager.md

Move all mibuploader/doc/*.png files to doc/img/ - Update the SNMP-Trap-Manager.md to reference these new images

mv mibuploader/src/etc/snmp/snmp.conf mibuploader/src/etc/snmp/snmptrapd.conf mibuploader/src/etc/snmp/snmptt.ini mibuploader/src/etc/sudoers to misc/ and ensure you update the docs to reference this location.

You shouldn't include files like you've done such as config.php, functions.php etc, you can use the loader we have:

$init_modules = array('polling', 'alerts');
require __DIR__ . '/includes/init.php';

As an example. You need to select the modules you might need. Most likely just webui.

We use composer for auto loading, you should make use of that where you can. The LibreNMS/ folder is where we store all of our OOP code at present.

Remove mibuploader/src/librenms/html/plugins/MIBUploader/test/phpunit-old.phar

Remove mibuploader/src/librenms/html/plugins/MIBUploader/test/test.sh

Any tests should use phpunit and be in tests/ with the relevant filename.

It looks like you are trying to include a templating style system to include various files within the html side of the webui. For now you should just have them included directly within the relevant files.

Move mibuploader/src/librenms/includes/snmptrap/genericTrap.inc.php to includes/snmptrap/

Move mibuploader/src/librenms/mibuploader/ssh/config to misc/

Move mibuploader/src/librenms/snmptrap.mibup.php to misc/

File: composer.json [hunk 1]
```diff
@@ -39,7 +38,8 @@
         "phpunit/phpunit": "5.*",
         "jakub-onderka/php-parallel-lint": "*",
         "jakub-onderka/php-console-highlighter": "*",
-        "fojuth/readmegen": "1.*"
+        "fojuth/readmegen": "1.*",
+        "amenadiel/jpgraph": "^3.6"
```
Comment (laf) [inline composer.json]: Any reason this is moved into the dev section?

File: misc/snmp.conf [hunk 2]
```diff
@@ -0,0 +1 @@
+mibdirs +/usr/share/snmp/mibs +/opt/librenms/mibs +/opt/librenms/mibuploader/snmpttmibs
```
Comment (laf) [inline misc/snmp.conf]: I'm not sure but /usr/share/snmp/mibs can probably cause issues so I'd remove that

File: misc/config [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# for fast testing, you should disable these options for production environments
```
Comment (laf) [inline misc/config]: What's this file for?

File: composer.json (hunk 1 above)
Comment (samyscoub) [reply composer.json]: Maybe added when I have tested the code, i will rollback to original content.

File: misc/snmp.conf (hunk 2 above)
Comment (samyscoub) [reply misc/snmp.conf]: Why this path causes a problem ?

File: misc/config (hunk 3 above)
Comment (samyscoub) [reply misc/config]: It is the `.ssh/config` sample file from documentation of the module inside this section :

```
10. Fill the `ssh/config` file from `mibuploader` directory and configure SSH commands into configuration. The SSH keys copy is mandatory.
```

File: composer.json (hunk 1 above)
Comment (samyscoub) [reply composer.json]: Yes, work in progress

**Comment by samyscoub** [comment]
@laf : how do you want to transform the content of `mibuploader/src/librenms/html/plugins/MIBUploader` to a core feature into `html/pages` ?
And where do you want to add a menu for this feature ? Into the `Alerts` menu ?

**Comment by laf** [comment]
> @laf : how do you want to transform the content of mibuploader/src/librenms/html/plugins/MIBUploader to a core feature into html/pages ?
> And where do you want to add a menu for this feature ? Into the Alerts menu ?

Yes if you can move to `html/pages/` that would be ace.

I'd say for now, it should be in the settings menu (the cog)

**Comment by laf** [comment]
How you getting on with this @samyscoub. Anything we can help with?

**Comment by samyscoub** [comment]
Hi @laf ,
I've started to refactor the code but not have the time to finish.
I can push my last changes and if somebody has time to finish my work ( change the import mode into the classes with your loader sample above ) ?
Thx !

**Comment by TheGreatDoc** [comment]
I really want to help with this, but I dont know where to start.

Any hint?

PS: Im not developer :(

**Comment by laf** [comment]
I'll be honest and say this will probably need someone who has some dev background, please don't take offence to that :)

**Comment by TheGreatDoc** [comment]
Dude, you hurt my feelings 🗡 

I can carry the coffes to the real devs.

So, is anyone stepping up with this? This is out of my dev skill but, if this is going to be mid-large time I can try to do some basic snmptrap handlers for, at least, bgp/authentication traps.

**Comment by laf** [comment]
@TheGreatDoc I don't drink coffee :)

I'm not aware of anyone who's going to get to it soon but afaik, it won't need specific handlers writing - it will support it all automagically

**Comment by TheGreatDoc** [comment]
@laf I mean while this is (or not) developed, to create basic handlers like the port up/down that comes with lnms.

Also, I can carry more drinks, not only coffe :)

**Comment by laf** [comment]
We already have port up / down snmptrap support :)

**Comment by samyscoub** [comment]
@TheGreatDoc , no problem for contributing, I've juste added you into contributors list from my forked project. Now, I have to add a submenu into the cog menu and change some things inside the plugin to match with @laf requirements. Once that work done, you will check the documentation of this feature to apply it in a real install case and make some tests. Is it possible for you to do this ?

**Comment by TheGreatDoc** [comment]
@samyscoub Sure!

**Comment by samyscoub** [comment]
@TheGreatDoc , can you try to validate this new feature with this release of my repo, inti Snmp trap manager feature from added documentation ?

**Comment by TheGreatDoc** [comment]
@samyscoub I'll do asap!

**Comment by DR3EVR8u8c** [comment]
Hello, 
this plugin is really what we need. I would like to have a test as well. However, when I run:
`./scripts/github-apply 5926`
get error messages:
```
Checking patch .github/PULL_REQUEST_TEMPLATE.md...
error: while searching for:

- [ ] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/) - please do NOT submit a pull request unless you have (signing the agreement in the same pull request is fine). Your commit message for signing the agreement must appear as per the docs.
- [ ] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

error: patch failed: .github/PULL_REQUEST_TEMPLATE.md:6
error: .github/PULL_REQUEST_TEMPLATE.md: patch does not apply
Checking patch scripts/github-apply...
error: scripts/github-apply: already exists in working directory
Checking patch scripts/github-remove...
error: scripts/github-remove: already exists in working directory
```
what I did wrong?

**Comment by DR3EVR8u8c** [comment]
does anyone work on this now? as mentioned above, I would like to test myself. However, is there any proper procedures to merge the changes to our test librenms. 
I ran the script which is described in beginning and failed. could anyone show me the correct command?

**Comment by laf** [comment]
I'll take a proper look when I get time but an initial scan shows that I think a lot of work is still needed to get this into shape for our code base.

**Comment by samyscoub** [comment]
Hi @laf, 
can you tell me what to change into my code to become compatible with your codebase ?
Thx !

**Comment by murrant** [comment]
There have been some updates to trap handling in LibreNMS recently: #8898 #9010 
The latter ads the ability to event log all traps and alert off of them.

Your code is very complex, which makes it hard to understand.  I tried to figured what happens when a trap is handled and it came down to just one `log_event()` call basically.  It would help if the code could be made easier to read and understand.

Specific tips:
 - Instead of including files directly, use PSR-4 autoloading.
 - Use the existing db schema management system, not your own.
 - Use the existing system for rendering web pages. (which is in the middle of a transition between legacy and Laravel)

**Comment by sippe2** [comment]
Any news about the project?

**Comment by samyscoub** [comment]
Any help to simplify the code ?

**Comment by murrant** [comment]
@samyscoub follow the process in your head and if there are any spots that are hard to follow see if you can make that better.  Removing layers of abstraction, if/else statements, removing/renaming temporary variables, and adding whitespace where appropriate are some of the things that help.

**Comment by laf** [comment]
I started to look at tidying up the code for this but it is too big a job for me right now.

**Comment by murrant** [comment]
If anyone wants to restart this project, please feel free to re-open/pull.
